File: src/game/generator.ts
/**
 * Level generator. Produces boards that are **guaranteed solvable** by shuffling a
 * balanced multiset of color segments and then verifying with the solver (rejection
 * sampling). Every returned level carries a concrete solution and its step count.
 *
 * A naive random-pour walk from the solved board does NOT guarantee solvability:
 * pouring same-color-onto-same-color merges runs and loses the split point, so the
 * walk isn't reversible. Verifying with a real solver sidesteps that entirely.
 */
import { isComplete, isWon } from './engine';
import { mulberry32 } from './rng';
import { bfsOptimal, canonical, solve } from './solver';
import type { Color, GameState, GeneratedLevel, Move, ParMode } from './types';

// Re-exported for callers that historically imported the PRNG from here; its home is now rng.ts.
export { mulberry32 };

/** Default palette ids (see ../theme/colors.ts). Generation uses the first N. */
export const PALETTE: readonly Color[] = [
  'ruby',
  'amethyst',
  'sapphire',
  'emerald',
  'amber',
  'rose',
  'teal',
  'violet',
  'lime',
  'tangerine',
  'cobalt',
  'magenta',
].map((id) => id as Color);

export const DEFAULT_CAPACITY = 4;
const MAX_COLORS = PALETTE.length; // 12
const MAX_RETRIES = 300;
/**
 * When a `minPar` floor is requested, stop *hunting for a harder board* after this many
 * solvable candidates and keep the most-tangled one seen. Bounds the extra solver work so
 * generation never churns — the floor is a preference, not a hard requirement.
 */
const PAR_SAMPLE_CAP = 50;

/**
 * Whether a (colors, bottles, capacity) combo is in the known-good, efficiently
 * solvable space. Requires at least one tube's worth of free space and caps colors at the
 * palette size. Note: with only 1 tube of slack, random layouts are frequently unsolvable,
 * so the generator's rejection sampling churns — callers should prefer >= 2 spare tubes for
 * fast, reliable generation (see the tier presets in levels.ts).
 */
export function isValidCombo(colors: number, bottles: number, capacity = DEFAULT_CAPACITY): boolean {
  if (!Number.isInteger(colors) || !Number.isInteger(bottles)) return false;
  if (capacity < 2) return false;
  if (colors < 2 || colors > MAX_COLORS) return false;
  const empties = bottles - colors;
  return empties >= 1;
}

/** Fisher–Yates shuffle using the provided PRNG (in place). */
function shuffle<T>(arr: T[], rng: () => number): T[] {
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    const tmp = arr[i]!;
    arr[i] = arr[j]!;
    arr[j] = tmp;
  }
  return arr;
}

/**
 * Build a randomized fill profile: how many segments each of `bottles` tubes starts with.
 * The fills sum to the total liquid (`colors * capacity`) and each sits in [0, capacity], so
 * the board still resolves into `colors` complete bottles.
 *
 * Unlike the old "full tubes plus trailing empties" layout, this scatters partially filled and
 * empty tubes for variety: a random number of tubes (0..empties) are left completely empty, and
 * whatever free space is left over is sprinkled across the remaining tubes as partial fills.
 * Picking the reserved-empty count freshly each attempt also keeps generation robust — when a
 * sparse, partial-heavy layout turns out unsolvable, a later attempt may reserve more empties and
 * fall back toward the easy classic shape. Tube positions are shuffled so empties land anywhere.
 */
function randomFillProfile(
  colors: number,
  bottles: number,
  capacity: number,
  rng: () => number,
): number[] {
  const empties = bottles - colors;
  const freeSpace = empties * capacity;

  // Reserve a random number of fully-empty tubes. Reserving more keeps the board easier and
  // generation more reliable; reserving fewer trades that for partial-tube variety.
  const reserved = Math.floor(rng() * (empties + 1));

  const fills = new Array<number>(bottles).fill(capacity);
  for (let i = 0; i < reserved; i++) fills[i] = 0;

  // Spread the leftover free space across the non-reserved tubes as partial fills, keeping each
  // of them at >= 1 segment so the empty-tube count stays exactly `reserved`.
  let toRemove = freeSpace - reserved * capacity;
  while (toRemove > 0) {
    const i = reserved + Math.floor(rng() * (bottles - reserved));
    if (fills[i]! > 1) {
      fills[i]!--;
      toRemove--;
    }
  }

  return shuffle(fills, rng);
}

/** Deal a flat list of segments into bottles sized by `fills` (segments must sum to the fills). */
function deal(segments: Color[], fills: number[], capacity: number): GameState {
  const bottles: Color[][] = [];
  let offset = 0;
  for (const fill of fills) {
    bottles.push(segments.slice(offset, offset + fill));
    offset += fill;
  }
  return { bottles, capacity };
}

/** A board is too easy if it's already won or has a pre-completed color bottle. */
function isDegenerate(state: GameState): boolean {
  if (isWon(state)) return true;
  return state.bottles.some((b) => b.length > 0 && isComplete(b, state.capacity));
}

export interface GenerateOptions {
  colors: number;
  bottles: number;
  capacity?: number;
  /** Seed for reproducibility. Defaults to a random seed. */
  seed?: number;
  /**
   * Reject boards easier than this par (best-of-N rejection sampling). The floor is a
   * preference: if no board reaches it within the sampling budget, the hardest one found is
   * returned anyway, so generation never fails on account of the floor. Omit for the legacy
   * "accept first solvable" behavior.
   */
  minPar?: number;
  /** How to measure par for the floor / the returned `par`. Defaults to `proxy`. */
  parMode?: ParMode;
}

/**
 * Measure a board's par. `optimal` runs the exact BFS (falling back to the DFS length if it
 * hits the node cap); `proxy` just uses the DFS solution length.
 */
function measurePar(state: GameState, solution: Move[], mode: ParMode): number {
  if (mode === 'optimal') {
    const optimal = bfsOptimal(state);
    if (optimal !== null) return optimal;
  }
  return solution.length;
}

/** The balanced multiset for a footprint: `capacity` copies of each of the first `colors` ids. */
function colorTemplate(colors: number, capacity: number): Color[] {
  const template: Color[] = [];
  for (let c = 0; c < colors; c++) {
    for (let k = 0; k < capacity; k++) template.push(PALETTE[c]!);
  }
  return template;
}

/**
 * One generation attempt: scatter the template into a random fill profile and verify solvability.
 * Returns the board and a proof solution, or `null` if the layout is degenerate or unsolvable.
 * Shared by `generateLevel` (rejection sampling) and `generateCandidates` (offline bake).
 */
function attemptBoard(
  template: Color[],
  colors: number,
  bottles: number,
  capacity: number,
  rng: () => number,
): { state: GameState; solution: Move[] } | null {
  const fills = randomFillProfile(colors, bottles, capacity, rng);
  const state = deal(shuffle([...template], rng), fills, capacity);
  if (isDegenerate(state)) return null;
  const solution = solve(state);
  if (!solution) return null;
  return { state, solution };
}

/**
 * Generate a verified-solvable level. Throws if the combo is invalid, or (extremely
 * unlikely) if no solvable board is found within the retry cap.
 */
export function generateLevel(options: GenerateOptions): GeneratedLevel {
  const capacity = options.capacity ?? DEFAULT_CAPACITY;
  const { colors, bottles } = options;

  if (!isValidCombo(colors, bottles, capacity)) {
    throw new Error(`Invalid combo: ${colors} colors / ${bottles} bottles / cap ${capacity}`);
  }

  const baseSeed = options.seed ?? (Math.random() * 2 ** 32) >>> 0;
  const parMode: ParMode = options.parMode ?? 'proxy';
  const minPar = options.minPar;
  const rng = mulberry32(baseSeed);
  const template = colorTemplate(colors, capacity);

  const build = (state: GameState, solution: Move[]): GeneratedLevel => ({
    state,
    colors,
    bottles,
    capacity,
    solution,
    minMoves: solution.length,
    par: measurePar(state, solution, parMode),
    seed: baseSeed,
  });

  // Best-of-N: the hardest solvable board seen so far, for the par-floor path's fallback.
  let best: GeneratedLevel | null = null;
  let solvableSeen = 0;

  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    const board = attemptBoard(template, colors, bottles, capacity, rng);
    if (!board) continue;
    const { state, solution } = board;

    // Legacy fast path: no floor requested, accept the first solvable board.
    if (minPar === undefined) return build(state, solution);

    const candidate = build(state, solution);
    if (best === null || candidate.par > best.par) best = candidate;
    if (candidate.par >= minPar) return candidate; // met the floor — done.

    // Give up hunting for a harder board once we've sampled enough; keep the best.
    if (++solvableSeen >= PAR_SAMPLE_CAP) return best;
  }

  if (best) return best;
  throw new Error(
    `Failed to generate a solvable ${colors}/${bottles} level after ${MAX_RETRIES} attempts`,
  );
}

/**
 * Collect up to `count` distinct, verified-solvable boards for a footprint — the candidate pool the
 * offline bake (`scripts/build-levels.ts`) scores and selects from. Unlike `generateLevel`, this
 * does not stop at the first acceptable board: it keeps sampling so a difficulty scorer can choose
 * the best fit for a target curve. Deduplicates by canonical board key, and bounds total work with
 * `maxAttempts` so an over-large `count` can never spin forever (returns however many it found).
 *
 * Par here is the cheap proxy (solution length); the bake computes its own exact difficulty metrics
 * per candidate, so paying for `optimal` par on every sampled board would be wasted work.
 */
export function generateCandidates(
  options: Pick<GenerateOptions, 'colors' | 'bottles' | 'capacity' | 'seed'>,
  count: number,
  maxAttempts = count * 40,
): GeneratedLevel[] {
  const capacity = options.capacity ?? DEFAULT_CAPACITY;
  const { colors, bottles } = options;
  if (!isValidCombo(colors, bottles, capacity)) {
    throw new Error(`Invalid combo: ${colors} colors / ${bottles} bottles / cap ${capacity}`);
  }

  const rng = mulberry32(options.seed ?? (Math.random() * 2 ** 32) >>> 0);
  const template = colorTemplate(colors, capacity);

  const seen = new Set<string>();
  const candidates: GeneratedLevel[] = [];
  for (let attempt = 0; attempt < maxAttempts && candidates.length < count; attempt++) {
    const board = attemptBoard(template, colors, bottles, capacity, rng);
    if (!board) continue;
    const key = canonical(board.state);
    if (seen.has(key)) continue;
    seen.add(key);
    candidates.push({
      state: board.state,
      colors,
      bottles,
      capacity,
      solution: board.solution,
      minMoves: board.solution.length,
      par: board.solution.length,
      seed: options.seed ?? 0,
    });
  }
  return candidates;
}

File: core/src/generator.rs
//! Level generator — port of `src/game/generator.ts`. RNG draw order is preserved
//! draw-for-draw (fill profile, its rejection loop, both shuffles, one stream across all
//! attempts), and the solver explores the identical node sequence, so a seed produces the
//! *identical board and solution* as JS — the strongest single differential check in F1.

use std::collections::HashSet;

use crate::engine::{is_complete, is_won};
use crate::rng::Mulberry32;
use crate::solver::{bfs_optimal, canonical, solve, DEFAULT_MAX_NODES};
use crate::state::{Key, State, Tube};
use crate::types::Move;

pub const DEFAULT_CAPACITY: u8 = 4;
pub const MAX_COLORS: usize = 12; // palette size
const MAX_RETRIES: usize = 300;
const PAR_SAMPLE_CAP: usize = 50;

#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub enum ParMode {
    Optimal,
    Proxy,
}

#[derive(Clone, Copy, Debug)]
pub struct GenerateOptions {
    pub colors: usize,
    pub bottles: usize,
    pub capacity: u8,
    pub seed: u32,
    /// Par floor for best-of-N rejection sampling; `None` = accept the first solvable board.
    pub min_par: Option<u32>,
    pub par_mode: ParMode,
}

#[derive(Clone, Debug)]
pub struct GeneratedLevel {
    pub state: State,
    pub colors: usize,
    pub bottles: usize,
    pub capacity: u8,
    pub solution: Vec<Move>,
    pub min_moves: usize,
    pub par: u32,
    pub seed: u32,
}

pub fn is_valid_combo(colors: usize, bottles: usize, capacity: u8) -> bool {
    if capacity < 2 {
        return false;
    }
    if !(2..=MAX_COLORS).contains(&colors) {
        return false;
    }
    bottles > colors // at least one empty tube
}

/// Fisher–Yates shuffle, one draw per swap from the back — identical to the JS helper.
fn shuffle<T>(arr: &mut [T], rng: &mut Mulberry32) {
    for i in (1..arr.len()).rev() {
        let j = (rng.next_f64() * (i + 1) as f64).floor() as usize;
        arr.swap(i, j);
    }
}

/// Randomized fill profile — port of `randomFillProfile`, including the reserved-empties
/// draw, the data-dependent sprinkle loop (a draw per iteration, applied only when the
/// picked tube can still shrink), and the final shuffle.
fn random_fill_profile(
    colors: usize,
    bottles: usize,
    capacity: u8,
    rng: &mut Mulberry32,
) -> Vec<usize> {
    let cap = capacity as usize;
    let empties = bottles - colors;
    let free_space = empties * cap;

    let reserved = (rng.next_f64() * (empties + 1) as f64).floor() as usize;

    let mut fills = vec![cap; bottles];
    for f in fills.iter_mut().take(reserved) {
        *f = 0;
    }

    let mut to_remove = free_space - reserved * cap;
    while to_remove > 0 {
        let i = reserved + (rng.next_f64() * (bottles - reserved) as f64).floor() as usize;
        if fills[i] > 1 {
            fills[i] -= 1;
            to_remove -= 1;
        }
    }

    shuffle(&mut fills, rng);
    fills
}

/// Deal a flat segment list into tubes sized by `fills`.
fn deal(segments: &[u8], fills: &[usize], capacity: u8) -> State {
    let mut tubes = Vec::with_capacity(fills.len());
    let mut offset = 0;
    for &fill in fills {
        tubes.push(Tube::from_cells(&segments[offset..offset + fill]));
        offset += fill;
    }
    State { tubes, capacity }
}

/// Already won, or a pre-completed color bottle.
fn is_degenerate(state: &State) -> bool {
    if is_won(state) {
        return true;
    }
    state
        .tubes
        .iter()
        .any(|t| !t.is_empty() && is_complete(t, state.capacity))
}

/// `capacity` copies of each of the first `colors` ids (ids ARE palette indices here).
fn color_template(colors: usize, capacity: u8) -> Vec<u8> {
    let mut template = Vec::with_capacity(colors * capacity as usize);
    for c in 0..colors as u8 {
        for _ in 0..capacity {
            template.push(c);
        }
    }
    template
}

fn measure_par(state: &State, solution: &[Move], mode: ParMode) -> u32 {
    if mode == ParMode::Optimal {
        if let Some(optimal) = bfs_optimal(state, None, DEFAULT_MAX_NODES) {
            return optimal as u32;
        }
    }
    solution.len() as u32
}

/// One generation attempt: profile, deal, degeneracy check, solve. RNG is consumed by the
/// profile + both shuffles BEFORE the degeneracy/solve rejection, so rejected attempts leave
/// the stream exactly where JS leaves it.
fn attempt_board(
    template: &[u8],
    colors: usize,
    bottles: usize,
    capacity: u8,
    rng: &mut Mulberry32,
) -> Option<(State, Vec<Move>)> {
    let fills = random_fill_profile(colors, bottles, capacity, rng);
    let mut segments = template.to_vec();
    shuffle(&mut segments, rng);
    let state = deal(&segments, &fills, capacity);
    if is_degenerate(&state) {
        return None;
    }
    let solution = solve(&state, None, DEFAULT_MAX_NODES)?;
    Some((state, solution))
}

/// Generate a verified-solvable level — port of `generateLevel` (legacy first-solvable path
/// and the best-of-N par-floor path).
pub fn generate_level(options: &GenerateOptions) -> Result<GeneratedLevel, String> {
    let GenerateOptions {
        colors,
        bottles,
        capacity,
        seed,
        min_par,
        par_mode,
    } = *options;
    if !is_valid_combo(colors, bottles, capacity) {
        return Err(format!(
            "invalid combo: {colors} colors / {bottles} bottles / cap {capacity}"
        ));
    }

    let mut rng = Mulberry32::new(seed);
    let template = color_template(colors, capacity);

    let build = |state: State, solution: Vec<Move>| -> GeneratedLevel {
        let par = measure_par(&state, &solution, par_mode);
        GeneratedLevel {
            min_moves: solution.len(),
            par,
            state,
            colors,
            bottles,
            capacity,
            solution,
            seed,
        }
    };

    let mut best: Option<GeneratedLevel> = None;
    let mut solvable_seen = 0usize;

    for _ in 0..MAX_RETRIES {
        let Some((state, solution)) = attempt_board(&template, colors, bottles, capacity, &mut rng)
        else {
            continue;
        };

        if min_par.is_none() {
            return Ok(build(state, solution));
        }

        let candidate = build(state, solution);
        let is_better = best.as_ref().is_none_or(|b| candidate.par > b.par);
        if candidate.par >= min_par.unwrap() {
            return Ok(candidate);
        }
        if is_better {
            best = Some(candidate);
        }

        solvable_seen += 1;
        if solvable_seen >= PAR_SAMPLE_CAP {
            return Ok(best.unwrap());
        }
    }

    best.ok_or_else(|| {
        format!(
            "failed to generate a solvable {colors}/{bottles} level after {MAX_RETRIES} attempts"
        )
    })
}

/// Up to `count` distinct solvable boards for the bake's candidate pool — port of
/// `generateCandidates` (dedupe by canonical key; proxy par).
pub fn generate_candidates(
    colors: usize,
    bottles: usize,
    capacity: u8,
    seed: u32,
    count: usize,
    max_attempts: usize,
) -> Result<Vec<GeneratedLevel>, String> {
    if !is_valid_combo(colors, bottles, capacity) {
        return Err(format!(
            "invalid combo: {colors} colors / {bottles} bottles / cap {capacity}"
        ));
    }

    let mut rng = Mulberry32::new(seed);
    let template = color_template(colors, capacity);

    let mut seen: HashSet<Key> = HashSet::new();
    let mut candidates: Vec<GeneratedLevel> = Vec::new();
    for _ in 0..max_attempts {
        if candidates.len() >= count {
            break;
        }
        let Some((state, solution)) = attempt_board(&template, colors, bottles, capacity, &mut rng)
        else {
            continue;
        };
        if !seen.insert(canonical(&state)) {
            continue;
        }
        candidates.push(GeneratedLevel {
            min_moves: solution.len(),
            par: solution.len() as u32,
            state,
            colors,
            bottles,
            capacity,
            solution,
            seed,
        });
    }
    Ok(candidates)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::engine::pour;

    #[test]
    fn generates_a_solvable_board_deterministically() {
        let opts = GenerateOptions {
            colors: 3,
            bottles: 5,
            capacity: 4,
            seed: 12345,
            min_par: None,
            par_mode: ParMode::Proxy,
        };
        let a = generate_level(&opts).unwrap();
        let b = generate_level(&opts).unwrap();
        assert_eq!(a.state, b.state);
        assert_eq!(a.solution, b.solution);

        // The stored solution actually wins.
        let mut cur = a.state.clone();
        for m in &a.solution {
            cur = pour(&cur, m.from as usize, m.to as usize, usize::MAX).0;
        }
        assert!(is_won(&cur));
    }

    #[test]
    fn candidates_are_distinct() {
        let out = generate_candidates(3, 5, 4, 999, 5, 200).unwrap();
        assert!(out.len() >= 2);
        let keys: HashSet<Key> = out.iter().map(|l| canonical(&l.state)).collect();
        assert_eq!(keys.len(), out.len());
    }
}
